website/reference/basic/configuration.md
---
layout: default
title: Configuration Guide
---

[SSLSecurity]: /mcollective/reference/plugins/security_ssl.html
[AESSecurity]: /mcollective/reference/plugins/security_aes.html
[Registration]: /mcollective/reference/plugins/registration.html
[Auditing]: /mcollective/simplerpc/auditing.html
[Authorization]: /mcollective/simplerpc/authorization.html
[Subcollectives]: /mcollective/reference/basic/subcollectives.html
[client_config]: /mcollective/configure/client.html
[server_config]: /mcollective/configure/server.html

> **Note:** For detailed information on MCollective's client and server daemon settings, see [Client Configuration Reference][client_config] and [Server Configuration Reference][server_config], respectively.

You can configure the MCollective server daemon and client by setting options in their configuration files. Plugins can also add their own options.

## Configuration Files

MCollective uses two configuration files, one for the client and one for the server.

If you don't specify a client configuration file when invoking MCollective, it looks for one in these locations, in the listed order:

1. `~/.mcollective`
1. `/etc/puppetlabs/mcollective/client.cfg`
1. `/etc/mcollective/client.cfg`

The MCollective server daemon looks in these locations, also in order:

1. `/etc/puppetlabs/agent/mcollective/server.cfg`
1. `/etc/mcollective/server.cfg`

The configuration file formats use a simple key-value syntax:

~~~ ini
key = value
~~~

## Common Options

|Key|Sample|Description|
|---|------|-----------|
|collectives|mcollective,subcollective|A list of [Subcollectives][] to join - 1.1.3 and newer only|
|main_collective|mcollective|The main collective to target - 1.1.3 and newer only|
|logfile|/var/log/mcollective.log|Where to log|
|loglevel|debug|Can be info, warn, debug, fatal, error|
|identity|dev1.your.com|Identifier for this node, does not need to be unique, defaults to hostname if unset and must match _/\w\.\-/_ if set|
|keeplogs|5|The amount of logs to keep|
|max_log_size|2097152|Max size in bytes for log files before rotation happens|
|libdir|/usr/libexec/mcollective:/site/mcollective|Where to look for plugins|
|connector|activemq|Which _connector_ plugin to use for communication|
|securityprovider|Psk|Which security model to use, see [SSL Security Plugin][SSLSecurity] and [AES Security Plugin][AESSecurity] for details on others|
|logger_type|file|Valid logger types, currently file, syslog or console|
|ssl_cipher|aes-256-cbc|This sets the cipher in use by the SSL code, see _man enc_ for a list supported by OpenSSL|
|direct_addressing|n|Enable or disable directed requests|
|direct_addressing_threshold|10|When direct requests are enabled, send direct messages for less than or equal to this many hosts|
|ttl|60|Sets the default TTL for requests - 1.3.2 and newer only|
|logfacility|When using the syslog logger sets the facility, defaults to user|
|default_discovery_method|The default method to use for discovery - _mc_ by default|
|default_discovery_options|Options to pass to the discovery plugin, empty by default|

## Server Configuration

The server configuration file should be readable by the root user _only._

|Key|Sample|Description|
|---|------|-----------|
|daemonize|1|Runs the server in the background|
|factsource|Facter|Which fact plugin to use|
|registration|Agentlist|[Registration][] plugin to use|
|registerinterval|120|How many seconds to sleep between registration messages, setting this to zero disables registration|
|registration_collective|development|Which sub-collective to send registration messages to|
|classesfile|/var/lib/puppet/classes.txt|Where to find a list of classes configured by your configuration management system|
|rpcaudit|1|Enables [SimpleRPC Auditing][Auditing]|
|rpcauditprovider|Logfile|Enables auditing using _MCollective::Audit::Logfile_|
|rpcauthorization|1|Enables [SimpleRPC Authorization][Authorization] globally|
|rpcauthprovider|action_policy|Use the _MCollective::Util::ActionPolicy_ plugin to manage authorization|
|rpclimitmethod|The method used for --limit-results. Can be either _first_ or _random_|
|fact_cache_time|300|How long to cache fact results for before refreshing from source|

## Client Configuration

The client configuration file should be globally readable.

> **Security Note:** _Don't_ put pre-shared keys or client SSL certificates in a world-readable file. See [Client Setup](#client-setup) for details on how to provide those values for each user.

|Key|Sample|Description|
|---|------|-----------|
|color|0|Disables the use of color in RPC results|
|connection_timeout|3|Sets the timeout for server communication. Default: none|
|discovery_timeout|2|Sets the timeout for discovering nodes. Default: 2|
|default_batch_size|100|Sets the default batch size which would be used when when not supplied in the API or CLI. Default 0|
|default_batch_sleep_time|30|Sets the default batch sleep time which would be used when when not supplied in the API or CLI. Default 1|

## Plugin Configuration

You can add configuration options for plugins you create, using this syntax:

~~~ ini
plugin.pluginname.option = value
~~~

Describe any options in the plugin's documentation.

Common plugin options include:

|Key|Sample|Description|
|---|------|-----------|
|plugin.yaml|/etc/puppetlabs/agent/mcollective/facts.yaml:/other/facts.yaml|Where the YAML fact source finds facts; multiples are merged|
|plugin.psk|123456789|The pre-shared key (PSK) to use for the PSK security provider|
|plugin.psk.callertype|group|What to base the callerid on for the PSK plugin: uid, gid, user, group, or identity|

## Client Setup

Do not set the host, user, password, and pre-shared key in the client configuration file, since these files are generally world-readable (unlike the server configuration file, which should be readable by the root user only).

> **Note:** You can make this clearer by explicitly setting these options to 'unset' in the client configuration file, which prevents MCollective from working unless something overrides those settings.

You can set per-user environment variables to supply these values:

~~~ bash
export STOMP_USER=user
export STOMP_PASSWORD=password
export MCOLLECTIVE_PSK=123456789
~~~

You can also set options that MCollective always applies by using the `MCOLLECTIVE_EXTRA_OPTS` environment variable:

~~~ bash
export MCOLLECTIVE_EXTRA_OPTS="--dt 5 --timeout 3 --config /home/you/mcollective.cfg"
~~~

The client library uses these variables, allowing you to give each administrative user their own username and privileges.

You can also configure clients in a user's `~/.mcollective` file as an alternative to the method above, but that file must then be a complete client configuration file; MCollective will not look for and apply other client configuration files after finding one at `~/.mcollective`.


website/changelog.md
---
layout: default
title: Changelog
toc: false
---

## Version 2.11.x

|Date|Description|Ticket|
|----|-----------|------|
|2017/09/20|Release *2.11.3*||
|2017/09/18|Speed up calculation of no responses and unexpected responses|MCO-818|
|2017/08/15|Release *2.11.2*||
|2017/08/02|Correctly determine reply status in all modes|MCO-815|
|2017/07/18|Release *2.11.1*||
|2017/06/28|Use OpenSSL::Cipher instead of OpenSSL::Cipher::Cipher to avoid warnings with Ruby 2.4|MCO-813|
|2017/06/28|Use Mutex instead of Thread.exclusive to avoid warnings with Ruby 2.4|MCO-812|
|2017/06/28|Use Integer instead of Fixnum to avoid warnings with Ruby 2.4|MCO-811|
|2017/06/21|Release *2.11.0*||
|2017/06/21|MCollective client now logs sending a request at INFO level with additional details, and server logs handling those requests at INFO level with similar details to help identify corresponding events.|MCO-784|
|2017/06/21|The default location for mcollective server logs is moved from `/var/log/puppetlabs` to `/var/log/puppetlabs/mcollective` (on Unix). No change has been made to Windows log locations.|MCO-783|
|2017/06/21|MCollective will consider symbols and their string representation equivalent for serializing/deserializing and accessing message keys as long as they're unambiguous.|MCO-799|
|2017/06/21|mcollective-client is now compatible with OpenSSL 1.1.0|MCO-804|
|2017/06/21|systemu has been upgraded to a more recent version that fixes an issue marshaling multi-byte characters.|MCO-806|


## Version 2.10.x

|Date|Description|Ticket|
|----|-----------|------|
|2017/06/09|Release *2.10.5*||
|2017/06/07|Fix acceptance source for JDK on Windows|MCO-808|
|2017/05/11|Release *2.10.4*||
|2017/04/06|Switch to using YAML.safe_load|MCO-794|
|2017/04/04|Release *2.10.3*||
|2017/03/30|Allow M::Client users to access M::Message|MCO-790|
|2017/03/09|Release *2.10.2*||
|2017/02/29|Release *2.10.1*||
|2017/02/06|Fix regression in --batch option|MCO-785|
|2017/01/23|Release *2.10.0*||
|2016/12/30|Updates for acceptance testing|MCO-782|
|2016/12/20|Fix broken link on plugin index page|DOCUMENT-618|
|2016/12/15|Fix `mco ping` breakage from prior commit in MCO-777|MCO-777|
|2016/12/08|Remove references to old PE versions||
|2016/11/21|Wait for all expected responses to rpc, report surprises|MCO-777|
|2016/11/16|Singletarget messaging|MCO-736|
|2016/11/07|Gemfile cleanup||
|2016/11/04|Use publish_timeout from config|MCO-778|
|2016/11/01|Support PQL in STDIN discovery|MCO-776|

## Version 2.9.x

|Date|Description|Ticket|
|----|-----------|------|
|2016/10/24|Release *2.9.1*||
|2016/09/23|Ping application includes discovery_timeout|MCO-775|
|2016/08/05|Release *2.9.0*||
|2016/06/24|Update Stomp gem dependency to >= 1.4.1|RE-7302|
|2016/06/13|acceptance: allow plugin versioned install||
|2016/06/01|Update README link to documentation||
|2016/05/24|Fix mco plugins installation when using puppet-agent on Debian family systems|MCO-688|
|2016/05/19|Move prepare_installation to run first|MCO-763|
|2016/05/17|Add --no-batch-files option|MCO-762|
|2016/05/17|Runtime username/password input to connect to middleware|MCO-760|

## Version 2.8.x

|Date|Description|Ticket|
|----|-----------|------|
|2016/06/27|Release *2.8.9*|MCO-761|
|2016/06/20|Make parsing of quoted strings in discovery filter expressions side effects free||
|2016/06/20|Do not use eval on unchecked strings in discovery filter expressions|MCO-765|
|2016/05/17|Link to PE docs|(#373)[https://github.com/puppetlabs/marionette-collective/pull/373]|
|2016/05/17|Update installation docs|(#376)[https://github.com/puppetlabs/marionette-collective/pull/376]|
|2016/05/11|Fix ActiveMQ install|(#378)[https://github.com/puppetlabs/marionette-collective/pull/378]|
|2016/05/03|Work around RuboCop parser change causing errors on non UTF-8 compliant strings|(#375)[https://github.com/puppetlabs/marionette-collective/pull/375]|
|2016/05/03|Pin rake to 10.4 required for running tests on ruby 1.8.7|(#375)[https://github.com/puppetlabs/marionette-collective/pull/375]|
|2016/02/25|Release *2.8.8*||
|2016/02/25|Fix dependency on aio for creating the pidfile directory|MCO-753|
|2016/02/18|Improve pidfile handling to avoid running multiple daemons|MCO-751|
|2016/02/18|Update build targets to LTS debian variants|MCO-634|
|2016/02/05|Update mco.bat file to allow for absence of CONFIG_FILE setting|MCO-748|
|2016/01/22|Add windows support to install.rb|MCO-745|
|2016/01/13|Release *2.8.7*||
|2015/12/17|Fix systemd logrotate unit|MCO-744|
|2015/12/16|Fix run helper for systems where the path to ruby binary includes spaces|MCO-742|
|2015/12/01|Fix negative data plugin comparisons|MCO-739|
|2015/09/15|Release *2.8.6*|MCO-726|
|2015/09/11|Fix solaris smf service manifest for aio|(#345)[https://github.com/puppetlabs/marionette-collective/pull/345]|
|2015/09/10|Release *2.8.5*|MCO-724|
|2015/08/25|Add condrestart to the suse init script for AIO|RE-11690|
|2015/08/21|Release *2.8.4*|MCO-706|
|2015/08/20|Revert RE-5032|MCO-705|
|2015/08/18|Release *2.8.3*|MCO-689|
|2015/08/05|Add solaris smf service for AIO|MCO-687|
|2015/07/17|Fully qualify the label in the osx mco plist|RE-5032|
|2015/07/13|Set character encoding in the OSX plist|(#326)[https://github.com/puppetlabs/marionette-collective/pull/326]|
|2015/07/03|Fix documentation links to use https|DOCS-2092|
|2015/06/05|Add `mco describe_filter` application|MCO-668|
|2015/06/05|Fix quote handling in compound query language|MCO-668|
|2015/06/04|Add acceptance tests|MCO-671|
|2015/05/19|Release *2.8.2*|MCO-648|
|2015/05/12|Do not build for debian stable or testing (target codenames instead)|MCO-665|
|2015/05/08|Exit non-zero when `mco` is called with a non-existent subcommand|MCO-640|
|2015/05/08|Add legacy `libdir` settings to aio sample config|MCO-641|
|2015/05/08|Downgrade warning on absent `libdir`|MCO-647|
|2015/05/04|Add OSX plist for AIO|MCO-646|
|2015/04/23|Drop lucid from build targets|MCO-638|
|2015/04/13|Drop fedora 19 from build targets|MCO-633|

lib/mcollective/config.rb
module MCollective
  # A pretty sucky config class, ripe for refactoring/improving
  class Config
    include Singleton

    attr_accessor :mode

    attr_reader :daemonize, :pluginconf, :configured
    attr_reader :logfile, :keeplogs, :max_log_size, :loglevel, :logfacility
    attr_reader :identity, :daemonize, :connector, :securityprovider, :factsource
    attr_reader :registration, :registerinterval, :classesfile
    attr_reader :rpcauditprovider, :rpcaudit, :configdir, :rpcauthprovider
    attr_reader :rpcauthorization, :color, :configfile
    attr_reader :rpclimitmethod, :logger_type, :fact_cache_time, :collectives
    attr_reader :main_collective, :ssl_cipher, :registration_collective
    attr_reader :direct_addressing, :direct_addressing_threshold, :ttl
    attr_reader :default_discovery_method, :default_discovery_options
    attr_reader :publish_timeout, :threaded, :soft_shutdown, :activate_agents
    attr_reader :registration_splay, :discovery_timeout, :soft_shutdown_timeout
    attr_reader :connection_timeout, :default_batch_size, :default_batch_sleep_time

    def initialize
      @configured = false
    end

    def loadconfig(configfile)
      set_config_defaults(configfile)

      if File.exists?(configfile)
        libdirs = []
        File.readlines(configfile).each do |line|

          # strip blank spaces, tabs etc off the end of all lines
          line.gsub!(/\s*$/, "")

          unless line =~ /^#|^$/
            if (line =~ /(.+?)\s*=\s*(.+)/)
              key = $1.strip
              val = $2

              begin
                case key
                when "registration"
                  @registration = val.capitalize
                when "registration_collective"
                  @registration_collective = val
                when "registerinterval"
                  @registerinterval = Integer(val)
                when "registration_splay"
                  @registration_splay = Util.str_to_bool(val)
                when "collectives"
                  @collectives = val.split(",").map {|c| c.strip}
                when "main_collective"
                  @main_collective = val
                when "logfile"
                  @logfile = val
                when "keeplogs"
                  @keeplogs = Integer(val)
                when "max_log_size"
                  @max_log_size = Integer(val)
                when "loglevel"
                  @loglevel = val
                when "logfacility"
                  @logfacility = val
                when "libdir"
                  paths = val.split(File::PATH_SEPARATOR)
                  paths.each do |path|
                    raise("libdir paths should be absolute paths but '%s' is relative" % path) unless Util.absolute_path?(path)

                    libdirs << path
                  end
                when "identity"
                  @identity = val
                when "direct_addressing"
                  @direct_addressing = Util.str_to_bool(val)
                when "direct_addressing_threshold"
                  @direct_addressing_threshold = Integer(val)
                when "color"
                  @color = Util.str_to_bool(val)
                when "daemonize"
                  @daemonize = Util.str_to_bool(val)
                when "securityprovider"
                  @securityprovider = val.capitalize
                when "factsource"
                  @factsource = val.capitalize
                when "connector"
                  @connector = val.capitalize
                when "classesfile"
                  @classesfile = val
                when /^plugin.(.+)$/
                  @pluginconf[$1] = val
                when "discovery_timeout"
                  @discovery_timeout = Integer(val)
                when "publish_timeout"
                  @publish_timeout = Integer(val)
                when "connection_timeout"
                  @connection_timeout = Integer(val)
                when "rpcaudit"
                  @rpcaudit = Util.str_to_bool(val)
                when "rpcauditprovider"
                  @rpcauditprovider = val.capitalize
                when "rpcauthorization"
                  @rpcauthorization = Util.str_to_bool(val)
                when "rpcauthprovider"
                  @rpcauthprovider = val.capitalize
                when "rpclimitmethod"
                  @rpclimitmethod = val.to_sym
                when "logger_type"
                  @logger_type = val
                when "fact_cache_time"
                  @fact_cache_time = Integer(val)
                when "ssl_cipher"
                  @ssl_cipher = val
                when "threaded"
                  @threaded = Util.str_to_bool(val)
                when "ttl"
                  @ttl = Integer(val)
                when "default_discovery_options"
                  @default_discovery_options << val
                when "default_discovery_method"
                  @default_discovery_method = val
                when "soft_shutdown"
                  @soft_shutdown = Util.str_to_bool(val)
                when "soft_shutdown_timeout"
                  @soft_shutdown_timeout = Integer(val)
                when "activate_agents"
                  @activate_agents = Util.str_to_bool(val)
                when "default_batch_size"
                  @default_batch_size = Integer(val)
                when "default_batch_sleep_time"
                  @default_batch_sleep_time = Float(val)
                when "topicprefix", "topicsep", "queueprefix", "rpchelptemplate", "helptemplatedir"
                  Log.warn("Use of deprecated '#{key}' option.  This option is ignored and should be removed from '#{configfile}'")
                else
                  raise("Unknown config parameter '#{key}'")
                end
              rescue ArgumentError => e
                raise "Could not parse value for configuration option '#{key}' with value '#{val}'"
              end
            end
          end
        end

        read_plugin_config_dir("#{@configdir}/plugin.d")

        raise 'Identities can only match /\w\.\-/' unless @identity.match(/^[\w\.\-]+$/)

        @configured = true

        libdirs.each do |dir|
          unless File.directory?(dir)
            Log.debug("Cannot find libdir: #{dir}")
          end

          # remove the old one if it exists, we're moving it to the front
          $LOAD_PATH.reject! { |elem| elem == dir }
          $LOAD_PATH.unshift dir
        end

        if @logger_type == "syslog"
          raise "The sylog logger is not usable on the Windows platform" if Util.windows?
        end

        PluginManager.loadclass("Mcollective::Facts::#{@factsource}_facts")
        PluginManager.loadclass("Mcollective::Connector::#{@connector}")
        PluginManager.loadclass("Mcollective::Security::#{@securityprovider}")
        PluginManager.loadclass("Mcollective::Registration::#{@registration}")
        PluginManager.loadclass("Mcollective::Audit::#{@rpcauditprovider}") if @rpcaudit
        PluginManager << {:type => "global_stats", :class => RunnerStats.new}

        Log.info("The Marionette Collective version #{MCollective::VERSION} started by #{$0} using config file #{configfile}")
      else
        raise("Cannot find config file '#{configfile}'")
      end
    end

    def set_config_defaults(configfile)
      @stomp = Hash.new
      @subscribe = Array.new
      @pluginconf = Hash.new
      @connector = "activemq"
      @securityprovider = "Psk"
      @factsource = "Yaml"
      @identity = Socket.gethostname
      @registration = "Agentlist"
      @registerinterval = 0
      @registration_collective = nil
      @registration_splay = false
      @classesfile = "/var/lib/puppet/state/classes.txt"
      @rpcaudit = false
      @rpcauditprovider = ""
      @rpcauthorization = false
      @rpcauthprovider = ""
      @configdir = File.dirname(configfile)
      @color = !Util.windows?
      @configfile = configfile
      @logger_type = "file"
      @keeplogs = 5
      @max_log_size = 2097152
      @rpclimitmethod = :first
      @fact_cache_time = 300
      @loglevel = "info"
      @logfacility = "user"
      @collectives = ["mcollective"]
      @main_collective = @collectives.first
      @ssl_cipher = "aes-256-cbc"
      @direct_addressing = true
      @direct_addressing_threshold = 10
      @default_discovery_method = "mc"
      @default_discovery_options = []
      @ttl = 60
      @mode = :client
      @publish_timeout = 2
      @threaded = false
      @soft_shutdown = false
      @soft_shutdown_timeout = nil
      @activate_agents = true
      @connection_timeout = nil
      @default_batch_size = 0
      @default_batch_sleep_time = 1
    end

    def libdir
      $LOAD_PATH
    end

    def read_plugin_config_dir(dir)
      return unless File.directory?(dir)

      Dir.new(dir).each do |pluginconfigfile|
        next unless pluginconfigfile =~ /^([\w]+).cfg$/

        plugin = $1
        File.open("#{dir}/#{pluginconfigfile}", "r").each do |line|
          # strip blank lines
          line.gsub!(/\s*$/, "")
          next if line =~ /^#|^$/
          if (line =~ /(.+?)\s*=\s*(.+)/)
            key = $1.strip
            val = $2
            @pluginconf["#{plugin}.#{key}"] = val
          end
        end
      end
    end
  end
end
